File: backend/app/api/v1/meeting_verification.py
from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.auth import AuthContext, CurrentUserDependency, require_kb_access, require_role
from app.core.exceptions import ConflictError, ForbiddenError, NotFoundError
from app.db.session import get_session
from app.models.meeting import AiTaskStatus, Meeting, MeetingQuestion, MeetingQuestionType
from app.schemas.analysis import AnalysisModuleOut, AnalysisRunRead, SourceRegistryItem
from app.schemas.kb import Role
from app.schemas.meeting import MeetingRead
from app.schemas.meeting_review import (
    AnalysisSelectionUpdate,
    AnalysisSubmissionRead,
    AnalysisSubmissionRequest,
    AnalysisTaskRead,
    MeetingQuestionCreate,
    MeetingQuestionRead,
    MeetingQuestionUpdate,
    MeetingVerificationRead,
    QuestionCandidatePage,
    QuestionCandidateRead,
    VerificationMutation,
)
from app.services.analysis_service import (
    get_analysis_task,
    get_latest_analysis_run,
    get_or_create_analysis_task,
    list_question_candidates,
    save_analysis_selection,
)
from app.services.meeting_review import MeetingReviewService

router = APIRouter(prefix="/meetings", tags=["会议核验"])
SessionDependency = Annotated[AsyncSession, Depends(get_session)]
EditorDependency = Annotated[AuthContext, Depends(require_role(Role.EDITOR))]


@router.get("/{meeting_id}/verification", response_model=MeetingVerificationRead)
async def get_verification(
    meeting_id: UUID, session: SessionDependency, current: CurrentUserDependency
) -> MeetingVerificationRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id)
    await _ensure_meeting_access(session, meeting, current, allow_deleted_kb=True)
    return await _snapshot(service, meeting)


@router.get(
    "/{meeting_id}/question-candidates",
    response_model=QuestionCandidatePage,
    summary="分页获取问题候选池",
)
async def list_candidates(
    meeting_id: UUID,
    session: SessionDependency,
    current: CurrentUserDependency,
    question_type: MeetingQuestionType = Query(...),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=5, ge=1, le=20),
) -> QuestionCandidatePage:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id)
    await _ensure_meeting_access(session, meeting, current)
    items, total = await list_question_candidates(
        session,
        meeting_id=meeting_id,
        question_type=question_type,
        offset=offset,
        limit=limit,
    )
    return QuestionCandidatePage(
        items=[
            QuestionCandidateRead(
                id=question.id,
                question_type=question.question_type,
                rank=question.candidate_rank,
                content=question.content,
                topic=question.topic,
                rationale=question.rationale,
                expected_answer_type=question.expected_answer_type,
                support_score=question.support_score,
                evidence_count=question.evidence_count,
                selected=question.analysis_selected,
                source=question.source,
                version=question.version,
            )
            for question in items
        ],
        total=total,
        offset=offset,
        limit=limit,
    )


@router.put(
    "/{meeting_id}/analysis-selection",
    response_model=MeetingVerificationRead,
    summary="保存带入 AI 分析的问题选择",
)
async def update_analysis_selection(
    meeting_id: UUID,
    payload: AnalysisSelectionUpdate,
    session: SessionDependency,
    current: EditorDependency,
) -> MeetingVerificationRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id, lock=True)
    await _ensure_meeting_access(session, meeting, current)
    if meeting.verification_version != payload.expected_version:
        raise ConflictError(
            "verification_version_conflict", "会议核验版本已变化，请刷新后重试"
        )
    await save_analysis_selection(
        session,
        meeting=meeting,
        selected_question_ids=payload.selected_question_ids,
        user_id=current.user_id,
    )
    meeting = await service.get_meeting(meeting_id)
    return await _snapshot(service, meeting)


@router.post(
    "/{meeting_id}/questions",
    response_model=MeetingQuestionRead,
    status_code=status.HTTP_201_CREATED,
)
async def create_question(
    meeting_id: UUID,
    payload: MeetingQuestionCreate,
    session: SessionDependency,
    current: EditorDependency,
) -> MeetingQuestionRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id, lock=True)
    await _ensure_meeting_access(session, meeting, current)
    question = await service.create_question(meeting_id, payload, user_id=current.user_id)
    return MeetingQuestionRead.model_validate(question)


@router.patch(
    "/{meeting_id}/questions/{question_id}", response_model=MeetingQuestionRead
)
async def update_question(
    meeting_id: UUID,
    question_id: UUID,
    payload: MeetingQuestionUpdate,
    session: SessionDependency,
    current: EditorDependency,
) -> MeetingQuestionRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id, lock=True)
    await _ensure_meeting_access(session, meeting, current)
    question = await service.update_question(
        meeting_id, question_id, payload, user_id=current.user_id
    )
    return MeetingQuestionRead.model_validate(question)


@router.delete(
    "/{meeting_id}/questions/{question_id}", status_code=status.HTTP_204_NO_CONTENT
)
async def delete_question(
    meeting_id: UUID,
    question_id: UUID,
    session: SessionDependency,
    current: EditorDependency,
    expected_version: int = Query(..., ge=1),
) -> Response:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id, lock=True)
    await _ensure_meeting_access(session, meeting, current)
    await service.delete_question(
        meeting_id,
        question_id,
        expected_version=expected_version,
        user_id=current.user_id,
    )
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{meeting_id}/verification/confirm", response_model=MeetingVerificationRead)
async def confirm_verification(
    meeting_id: UUID,
    payload: VerificationMutation,
    session: SessionDependency,
    current: EditorDependency,
) -> MeetingVerificationRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id, lock=True)
    await _ensure_meeting_access(session, meeting, current)
    meeting = await service.confirm(
        meeting_id, expected_version=payload.expected_version, user_id=current.user_id
    )
    return await _snapshot(service, meeting)


@router.post("/{meeting_id}/analysis-submissions", response_model=AnalysisSubmissionRead)
async def submit_analysis(
    meeting_id: UUID,
    payload: AnalysisSubmissionRequest,
    session: SessionDependency,
    current: EditorDependency,
) -> AnalysisSubmissionRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id, lock=True)
    await _ensure_meeting_access(session, meeting, current)
    task = await get_or_create_analysis_task(
        session,
        meeting=meeting,
        organization_id=current.organization_id,
        expected_version=payload.expected_version,
        selected_question_ids=payload.selected_question_ids,
        user_id=current.user_id,
    )
    meeting = await service.get_meeting(meeting_id)
    return AnalysisSubmissionRead(
        verification=await _snapshot(service, meeting),
        message="AI 分析已提交，生成完成后可在 AI 纪要分析页查看结果。",
        task_id=task.id,
        task_status=str(
            task.status.value if isinstance(task.status, AiTaskStatus) else task.status
        ),
    )


@router.get("/{meeting_id}/analysis-task", response_model=AnalysisTaskRead)
async def get_analysis_task_status(
    meeting_id: UUID,
    session: SessionDependency,
    current: CurrentUserDependency,
) -> AnalysisTaskRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id)
    await _ensure_meeting_access(session, meeting, current)
    task = await get_analysis_task(session, meeting_id)
    if task is None:
        raise NotFoundError("AI 分析任务", "analysis_task_not_found")
    return AnalysisTaskRead(
        task_id=task.id,
        status=str(task.status.value if isinstance(task.status, AiTaskStatus) else task.status),
        current_stage=task.current_stage,
        progress=task.progress,
        message=task.message,
        error_message=task.error_message,
        retry_count=task.retry_count,
    )


@router.post("/{meeting_id}/analysis/reanalyze", response_model=AnalysisTaskRead)
async def reanalyze_meeting(
    meeting_id: UUID,
    payload: AnalysisSubmissionRequest,
    session: SessionDependency,
    current: EditorDependency,
) -> AnalysisTaskRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id, lock=True)
    await _ensure_meeting_access(session, meeting, current)
    existing = await get_analysis_task(session, meeting_id)
    if existing is None:
        raise NotFoundError("AI 分析任务", "analysis_task_not_found")
    selected_ids = list(
        (
            await session.scalars(
                select(MeetingQuestion.id).where(
                    MeetingQuestion.meeting_id == meeting_id,
                    MeetingQuestion.deleted_at.is_(None),
                    MeetingQuestion.analysis_selected.is_(True),
                )
            )
        ).all()
    )
    task = await get_or_create_analysis_task(
        session,
        meeting=meeting,
        organization_id=current.organization_id,
        expected_version=payload.expected_version,
        selected_question_ids=selected_ids,
        user_id=current.user_id,
        force_reanalyze=True,
    )
    return AnalysisTaskRead(
        task_id=task.id,
        status=str(task.status.value if isinstance(task.status, AiTaskStatus) else task.status),
        current_stage=task.current_stage,
        progress=task.progress,
        message=task.message,
        error_message=task.error_message,
        retry_count=task.retry_count,
    )


@router.get("/{meeting_id}/analysis", response_model=AnalysisRunRead)
async def get_analysis_result(
    meeting_id: UUID,
    session: SessionDependency,
    current: CurrentUserDependency,
) -> AnalysisRunRead:
    service = MeetingReviewService(session)
    meeting = await service.get_meeting(meeting_id)
    await _ensure_meeting_access(session, meeting, current)
    run = await get_latest_analysis_run(session, meeting_id)
    if run is None:
        raise NotFoundError("AI 分析结果", "analysis_result_not_found")
    return AnalysisRunRead(
        task_id=run.task_id,
        source_version=run.source_version,
        modules=[AnalysisModuleOut.model_validate(module) for module in run.modules],
        sources=[SourceRegistryItem.model_validate(source) for source in run.sources],
        insufficient_notes=[str(note) for note in run.insufficient_notes],
        created_at=run.created_at.isoformat(),
    )


async def _snapshot(
    service: MeetingReviewService, meeting: Meeting
) -> MeetingVerificationRead:
    cut_points, open_ended = await service.list_questions(meeting.id)
    return MeetingVerificationRead(
        meeting=MeetingRead.model_validate(meeting),
        cut_point_questions=[MeetingQuestionRead.model_validate(q) for q in cut_points],
        open_ended_questions=[MeetingQuestionRead.model_validate(q) for q in open_ended],
        verification_version=meeting.verification_version,
        eligibility=service.eligibility(meeting, cut_points, open_ended),
    )


async def _ensure_meeting_access(
    session: AsyncSession,
    meeting: Meeting,
    current: AuthContext,
    *,
    allow_deleted_kb: bool = False,
) -> None:
    if meeting.organization_id != current.organization_id:
        raise ForbiddenError("会议不存在或无权访问")
    if meeting.knowledge_base_id is not None and not allow_deleted_kb:
        await require_kb_access(session, current, meeting.knowledge_base_id)

File: frontend/src/api/meetingVerification.ts
import { http } from '@/api/client'
import { normalizeQuestionEvidence, normalizeQuestionGenerationTask, normalizeVerification } from '@/types/meetingVerification'
export { normalizeVerification } from '@/types/meetingVerification'
import { toApiError } from '@/utils/errors'
import type {
  AnalysisSelectionPayload,
  AnalysisTask,
  AnalysisSubmissionResponse,
  MeetingVerificationSnapshot,
  QuestionCandidate,
  QuestionCandidatePage,
  VerificationActionPayload,
  VerificationQuestion,
  VerificationQuestionCreatePayload,
  VerificationQuestionUpdatePayload,
  QuestionEvidence,
  QuestionGenerationTask,
} from '@/types/meetingVerification'

export function normalizeCandidate(raw: unknown): QuestionCandidate | null {
  if (!raw || typeof raw !== 'object') return null
  const item = raw as Record<string, unknown>
  const id = String(item.id ?? item.question_id ?? '')
  const content = typeof item.content === 'string' ? item.content.trim() : ''
  if (!id || !content) return null
  const type = item.question_type === 'open_ended' || item.question_type === 'OPEN_ENDED' ? 'open_ended' : 'cut_point'
  return {
    id,
    question_type: type,
    rank: typeof item.rank === 'number' ? item.rank : (typeof item.candidate_rank === 'number' ? item.candidate_rank : null),
    content,
    topic: typeof item.topic === 'string' ? item.topic : null,
    rationale: typeof item.rationale === 'string' ? item.rationale : null,
    expected_answer_type: typeof item.expected_answer_type === 'string' ? item.expected_answer_type : null,
    support_score: typeof item.support_score === 'number' ? item.support_score : null,
    evidence_count: typeof item.evidence_count === 'number' ? item.evidence_count : 0,
    selected: item.selected === true || item.analysis_selected === true,
    source: typeof item.source === 'string' ? item.source : 'ai',
    version: typeof item.version === 'number' ? item.version : 1,
  }
}

export function normalizeAnalysisTask(raw: unknown): AnalysisTask | null {
  if (!raw || typeof raw !== 'object') return null
  const item = raw as Record<string, unknown>
  const taskId = String(item.task_id ?? item.taskId ?? '')
  if (!taskId) return null
  return {
    task_id: taskId,
    status: String(item.status ?? 'QUEUED'),
    current_stage: String(item.current_stage ?? item.currentStage ?? 'queued'),
    progress: Number(item.progress ?? 0) || 0,
    message: typeof item.message === 'string' ? item.message : null,
    error_message: typeof item.error_message === 'string' ? item.error_message : null,
    retry_count: Number(item.retry_count ?? item.retryCount ?? 0) || 0,
  }
}

const endpoint = (meetingId: string) => `/api/v1/meetings/${meetingId}/verification`
const questionsEndpoint = (meetingId: string) => `/api/v1/meetings/${meetingId}/questions`
const questionGenerationEndpoint = (meetingId: string) => `/api/v1/meetings/${meetingId}/question-generation`

export const meetingVerificationApi = {
  async get(meetingId: string): Promise<MeetingVerificationSnapshot> {
    const { data } = await http.get(endpoint(meetingId))
    return normalizeVerification(data)
  },
  async createQuestion(meetingId: string, payload: VerificationQuestionCreatePayload): Promise<VerificationQuestion> {
    const { data } = await http.post(questionsEndpoint(meetingId), payload)
    return data
  },
  async updateQuestion(meetingId: string, questionId: string, payload: VerificationQuestionUpdatePayload): Promise<VerificationQuestion> {
    const { data } = await http.patch(`${questionsEndpoint(meetingId)}/${questionId}`, payload)
    return data
  },
  async removeQuestion(meetingId: string, questionId: string, expected_version: number): Promise<void> {
    await http.delete(`${questionsEndpoint(meetingId)}/${questionId}`, { params: { expected_version } })
  },
  async confirm(meetingId: string, payload: VerificationActionPayload): Promise<MeetingVerificationSnapshot> {
    const { data } = await http.post(`${endpoint(meetingId)}/confirm`, payload)
    const normalized = normalizeVerification(data?.verification ?? data)
    return normalized.meeting?.id ? normalized : { ...normalized, meeting_id: normalized.meeting_id ?? meetingId }
  },
  async submitAnalysis(meetingId: string, payload: AnalysisSelectionPayload): Promise<AnalysisSubmissionResponse> {
    const { data } = await http.post(`/api/v1/meetings/${meetingId}/analysis-submissions`, payload)
    return {
      verification: normalizeVerification(data?.verification ?? data),
      message: typeof data?.message === 'string' ? data.message : 'AI 分析已提交。',
      task_id: typeof data?.task_id === 'string' ? data.task_id : null,
      task_status: typeof data?.task_status === 'string' ? data.task_status : null,
    }
  },
  async getQuestionCandidates(meetingId: string, questionType: 'cut_point' | 'open_ended', offset = 0, limit = 5): Promise<QuestionCandidatePage> {
    const { data } = await http.get<unknown>(`/api/v1/meetings/${meetingId}/question-candidates`, {
      params: { question_type: questionType, offset, limit },
    })
    const value = (data && typeof data === 'object' ? data : {}) as Record<string, unknown>
    return {
      items: (Array.isArray(value.items) ? value.items : []).map(normalizeCandidate).filter((item): item is QuestionCandidate => Boolean(item)),
      total: Number(value.total ?? 0) || 0,
      offset: Number(value.offset ?? offset) || 0,
      limit: Number(value.limit ?? limit) || limit,
    }
  },
  async saveAnalysisSelection(meetingId: string, payload: AnalysisSelectionPayload): Promise<MeetingVerificationSnapshot> {
    const { data } = await http.put(`/api/v1/meetings/${meetingId}/analysis-selection`, payload)
    return normalizeVerification(data)
  },
  async getAnalysisTask(meetingId: string): Promise<AnalysisTask | null> {
    try {
      const { data } = await http.get(`/api/v1/meetings/${meetingId}/analysis-task`)
      return normalizeAnalysisTask(data)
    } catch (error) {
      if (toApiError(error).status === 404) return null
      throw error
    }
  },
  async reanalyzeAnalysis(meetingId: string, payload: AnalysisSelectionPayload): Promise<AnalysisTask> {
    const { data } = await http.post(`/api/v1/meetings/${meetingId}/analysis/reanalyze`, payload)
    const task = normalizeAnalysisTask(data)
    if (!task) throw new Error('重新分析任务格式不正确。')
    return task
  },
  async getQuestionGeneration(meetingId: string): Promise<QuestionGenerationTask | null> {
    try {
      const { data } = await http.get(questionGenerationEndpoint(meetingId))
      return normalizeQuestionGenerationTask(data)
    } catch (error) {
      // Older meetings were created before question-generation tasks existed.
      if (toApiError(error).status === 404) return null
      throw error
    }
  },
  async retryQuestionGeneration(meetingId: string): Promise<QuestionGenerationTask> {
    const { data } = await http.post(`${questionGenerationEndpoint(meetingId)}/retry`)
    return normalizeQuestionGenerationTask(data)
  },
  async getQuestionEvidences(meetingId: string, questionId: string): Promise<QuestionEvidence[]> {
    try {
      const { data } = await http.get(`${questionsEndpoint(meetingId)}/${questionId}/evidences`)
      const values = Array.isArray(data) ? data : Array.isArray(data?.evidences) ? data.evidences : Array.isArray(data?.items) ? data.items : []
      return values.map(normalizeQuestionEvidence)
    } catch (error) {
      if (toApiError(error).status === 404) return []
      throw error
    }
  },
}
